tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+			existingNtfn.proofsRegistrations, newProofs,
+		)
+
+		return nil
+	}
+
+	reOrgChan := make(chan struct{}, 1)
+	ctxc, cancel := context.WithCancel(ctx)
+
+	// We only register for a single confirmation, which we expect to come
+	// in immediately, since we're only given the proof after one conf.
+	// But we keep the registration open, so we will get notified again if
+	// there is a re-org.
+	confEvent, errChan, err := w.cfg.ChainBridge.RegisterConfirmationsNtfn(
+		ctxc, &txHash, tx.TxOut[0].PkScript, 1, firstProof.BlockHeight,
+		true, reOrgChan,
+	)
+	if err != nil {
+		cancel()
+		return fmt.Errorf("unable to register for conf ntfn: %w", err)
+	}
+
+	// Now that we have created the confirmation watcher, we can add the
+	// new proofs to the pending proofs map.
+	w.pendingProofs[txHash] = &anchorTxNotification{
+		proofsRegistrations: []*proofRegistration{newProofs},
+		cancel:              cancel,
+	}
+
+	w.Wg.Add(1)
+	go func() {
+		defer confEvent.Cancel()
+		defer w.Wg.Done()
+		defer cancel()
+
+		for {
+			select {
+			// If the transaction was confirmed normally, we don't
+			// need to do anything and can stop watching it.
+			case conf, ok := <-confEvent.Confirmed:
+				if !ok {
+					return
+				}
+
+				select {
+				case w.incomingConfs <- conf:
+				case <-w.Quit:
+				}
+
+			// If the transaction was re-organized out, we need to
+			// update the proof and store it in the proof archive.
+			case <-reOrgChan:
+				log.Infof("Anchor TX %v was re-organized out "+
+					"of the chain, will update proof "+
+					"with next confirmation", txHash)
+
+				// We continue to watch the transaction until
+				// it reaches a safe confirmation depth. We
+				// basically only use this signal to log the
+				// message above. We expect another confirmation
+				// to come in once the transaction is included
+				// in a new block in the re-organized chain.
+
+			case err := <-errChan:
+				if !fn.IsCanceled(err) {
+					w.reportErr(fmt.Errorf("error while "+
+						"waiting for conf: %w", err))
+					return
+				}
+
+			case <-ctx.Done():
+				if !fn.IsCanceled(ctx.Err()) {
+					log.Warnf("Stopping to watch TX %v "+
+						"due to context error: %v",
+						txHash, ctx.Err())
+				}
+
+				return
+
+			case <-w.Quit:
+				log.Debugf("Stopping to watch TX %v, re-org "+
+					"watcher shutting down", txHash)
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// updateProofs updates the given proofs with the new block and merkle proof and
+// then informs the caller about the update.
+func (w *ReOrgWatcher) updateProofs(proofNtfn *anchorTxNotification,
+	conf *chainntnfs.TxConfirmation) error {
+
+	for idxR := range proofNtfn.proofsRegistrations {
+		r := proofNtfn.proofsRegistrations[idxR]
+		for idxP := range r.proofs {
+			p := r.proofs[idxP]
+
+			// We can now update the proof with the new
+			// block.
+			p.BlockHeight = conf.BlockHeight
+			p.BlockHeader = conf.Block.Header
+			merkleProof, err := proof.NewTxMerkleProof(
```

Comment:
If all the proofs in a registration share an anchor TX, I think we can make this merkle proof once per-registration and then just copy it into each actual asset proof?